### PR TITLE
add command for usages out of the editor

### PR DESCRIPTION
**Pull request by YueHui**

### CODE REVIEW

**Comment by antoniolucasnobar** [comment]
Hello @YueHui !

First of all, thanks for your contribution. Create a command is a must for a ckeditor plugin!

I made some minor adjustments, removed some duplicated code and some code formatting. If you please could see if it is everything working fine in your environment, it would be great.

I let one or two notes in the code, if you please could look into, I appreciate.

File: src/findReplace.js [hunk 1]
```diff
@@ -219,147 +220,39 @@ export default class FindReplace extends Plugin {
...
-		}
+	_replace( findField, replaceField,replaceAll=false ) {
+		const findText = findField.fieldView.element.value;
+		const replaceText = replaceField.fieldView.element.value;
+		this.editor.execute('findReplace',{
+			findText,
+			replaceText,
+			replaceAll
+		})
+		this._find(findField,0)
```
Comment (antoniolucasnobar) [inline src/findReplace.js]: @YueHui do we really need to call `this._find` here? It seems it is already called inside replace method in FindCommand. Could you please check if it is really necessary?

**Review by antoniolucasnobar — COMMENTED** [review]
@YueHui I like how you created the command, very flexible. It was clever!

But I do not know if it should be only one command or two (one for find and one for replace text). It feels a little strange to have only one command. 

So, I would like to know your opinion.

File: src/findReplace.js (hunk 1 above)
Comment (YueHui) [reply src/findReplace.js]: the purpose of call `this._find` is to update the `findField` infotext,it's Inappropriate, i fixed it

**Comment by YueHui** [comment]
if split find and replace command the replace command still needs findText and replaceText parameters,it's not much more convenient for the user,only just a name difference

**Review by antoniolucasnobar — CHANGES_REQUESTED** [review]
Hi @YueHui , I finally got to test this on my environment. Sorry the delay.

I fixed the `Command` import to comply with other imports in the project. I think it will not impact your use, but, please let me know.

I noted three things using the editor interface, different from version 0.1.1:
- If I try to replace all occurrences, no message is shown to the user;
- If I try to replace a text without previously search it, it performs just a search without replace. I have to click again the replace button to actually replace the text. The same occurs if I try to replace out of editor (using `editor.execute...`)
- The console throws an error every time the code does a `this.editor.execute` when running against ckeditor5 v20; But everything works. This could be a bug in this ckeditor version and it is not really important right now. I set the compatibility to >=v21, so users of the plugin should be aware it is not recomended to use it with previous versions.

Could you please check at least the two first items, so we can fix it, before merge this pull request? 

thanks in advance.

File: src/findReplace.js [hunk 2]
```diff
@@ -219,147 +220,43 @@ export default class FindReplace extends Plugin {
...
-		const replaceBy = replaceField.fieldView.element.value;
-		model.change( writer => {
-			const markers = model.markers.getMarkersGroup( SEARCH_MARKER );
-			let size = 0;
-			for ( const marker of markers ) {
-				model.insertContent( writer.createText( replaceBy ), marker.getRange() );
-				size++;
-			}
-			this._resetStatus();
-			replaceField.infoText = t( 'Replaced ' ) + size + t( ' times' );
```
Comment (antoniolucasnobar) [inline src/findReplace.js:344]: I think this was lost in the new version. Without it, the user get no feedback. the replace method could return total number of replacements (in the normal use, would be always 1, but in the replaceAll version, it should be variable)

File: src/findCommand.js [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+			currentIndex:this.currentSearchIndex,
+			total: markers.length,
+		};
+    }
+
+    _replace(findText,replaceText ) {
+		const model = this.editor.model;
+		const markers = Array.from( model.markers.getMarkersGroup( SEARCH_MARKER ) );
+		const sameSearch = isSameSearch( findText, markers );
+		const currentMarker = sameSearch ? markers[ this.currentSearchIndex ] : this._find( findText, 1 );
```
Comment (antoniolucasnobar) [inline src/findCommand.js]: maybe the problem with replace not being fired is here, now that `_find` returns an object, not only the currentMarker. I think destructure may fix this.

**Comment by antoniolucasnobar** [comment]
Hello @YueHui 

I think I fixed the problems above. Could you please test in your environment to check if everything is ok? I do not want to make a release which may not work in your environment.

About the code, I am not so happy how I did it, but it is working. Maybe you can provide a more elegant solution. 

One last thing. I plan to support case sensitive search in the future (issue #3), so, as you provided an zh-cn translation, could you please add an translation to this feature? I added a "Match case" entry in contexts.json.

One more time, thank you for your help in this project.

**Comment by YueHui** [comment]
sorry for reply so late, i didn't open computer at weekend,it works fine in my environment ,but when replace a text without search it, it directly replace the first item and mark the second item ,is it feel a little kind of strange?i think the origin logic is fine ,first search it ,and then the user decide to replace or not

**Comment by antoniolucasnobar** [comment]
> sorry for reply so late, i didn't open computer at weekend,it works fine in my environment ,but when replace a text without search it, it directly replace the first item and mark the second item ,is it feel a little kind of strange?i think the origin logic is fine ,first search it ,and then the user decide to replace or not

Well, I think if the user directly clicks on the replace option, in my opinion, it should replace the first occurrence. the plugin has next and previous buttons to search (maybe we need better names for them, Find Next and Find Previous).

I made some tests in Google Docs and in LibreOffice Writer. 

Google docs search as you type, so the replace works from the first hit (this was what I had in mind when I created the plugin, but I still did not test the search as you type (I do not know if it can slow down the editor).

LibreOffice Writer behaves like you described. If I open the replace dialog (CTRL+H) and hit replace button, the first time, it performs only a search for the first item. If I hit again, the it replaces the first occurrence.
I noted a difference only if I CTRL+H with some text selected, then click the replace button, it performs the replace the first time (it is like the first search was already made).

I am going to think in this behavior better tomorow, it is a little late here in Brazil.

**Review by antoniolucasnobar — APPROVED** [review]

**Comment by antoniolucasnobar** [comment]
Hi @YueHui I changed behavior. If replace is triggered before search, only search is made. same behavior as Libreoffice and accordingly to your vision.

I fixed a problem with negative numbers in find (if the negative number is bigger than the number of occurrences).

I added support to replace to accept the increment argument:

- this allows to the user in the interface to use shift+enter to replace backwards
- this also allows to use the increment parameter in findReplaceCommand to replace and choose the next item to be replaced

Maybe I should support only backward or forward search/replace (increment or decrement by 1). But this model makes it more flexible.

I will release 0.2.0; If it does not work fine in your environment, please let me know (with maybe another pull request). 

thanks for your contribution!

**Review by antoniolucasnobar — APPROVED** [review]

**Pull request merged**